### PR TITLE
Test C++ out-of-line operator functions

### PR DESCRIPTION
**Pull request by MForster**
Add a test to verify that C++ out-of-line operator functions are imported correctly.

This is part of addressing SR-12748.

### CODE REVIEW

**Comment by MForster** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
Could you add SILGen and IRGen tests for just one of the operators?

File: test/Interop/Cxx/operators/non-member-out-of-line.swift [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+OperatorsTestSuite.test("slash") {
+  let lhs = IntBox(value: 42)
+  let rhs = IntBox(value: 23)
+
+  let result = lhs / rhs
+
+  expectEqual(1, result.value)
+}
```
Comment (martinboehme) [inline test/Interop/Cxx/operators/non-member-out-of-line.swift:48]: The tests seem a bit redundant -- do we really need all four of these operators?

Why would a single operator not be enough -- in other words, what's the difference in test coverage between the tests?

**Comment by MForster** [comment]
> Could you add SILGen and IRGen tests for just one of the operators? 

Done.

File: test/Interop/Cxx/operators/non-member-out-of-line.swift (hunk 1 above)
Comment (MForster) [reply test/Interop/Cxx/operators/non-member-out-of-line.swift:48]: Well, yes, I deliberated that a bit as well. Happy to drop if you prefer.

The background is that there are quite a few different [operators](https://github.com/apple/llvm-project/blob/swift/master/clang/include/clang/Basic/OperatorKinds.def#L58). Some of them are unary, some binary. Some are allowed as member functions, some are not. Some we will import in different ways (`operator[]` as subscript), some not at all.

So, there will be code that treats different operators differently. Of course not these four. But there will be different code branches. And I want to be sure that every imported operator gets treated right. I guess that the inline version of the tests may sufficient. I added this as a bit extra precaution.

Wdyt?

Comment (martinboehme) [reply test/Interop/Cxx/operators/non-member-out-of-line.swift:48]: That makes sense in principle, but I believe the four operators you've got here are all treated the same from the point of import -- right?

I would suggest doing one of these:
*    Include just `operator+` and rely on the inline version of the test for coverage of operators that import differently.
*    Include multiple operators, but choose ones that import differently. If this takes a lot of code, it's probably not worth doing though.

**Review by gribozavr — APPROVED** [review]

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by MForster** [comment]
@swift-ci Please tests

**Comment by MForster** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@swift-ci Please test Windows

**Comment by gribozavr** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
@MForster Please update the tests to pass with MSVC mangled names on Windows.

**Comment by MForster** [comment]
@swift-ci Please test

**Review by martinboehme — APPROVED** [review]

**Comment by MForster** [comment]
@swift-ci Please test Windows

**Comment by MForster** [comment]
@swift-ci Please test Windows

File: test/Interop/Cxx/operators/non-member-out-of-line-irgen.swift [hunk 2]
```diff
@@ -4,5 +4,5 @@ import NonMemberOutOfLine
 
 public func add(_ lhs: IntBox, _ rhs: IntBox) -> IntBox { lhs + rhs }
 
-// CHECK: call i32 @_Zpl6IntBoxS_(i32 %{{[0-9]+}}, i32 %{{[0-9]+}})
-// CHECK: declare i32 @_Zpl6IntBoxS_(i32, i32)
+// CHECK: call i32 [[NAME:@.*]](i32 %{{[0-9]+}}, i32 %{{[0-9]+}})
+// CHECK: declare i32 [[NAME]](i32, i32)
```
Comment (gribozavr) [inline test/Interop/Cxx/operators/non-member-out-of-line-irgen.swift:8]: This makes it a rather generic test -- I'd rather prefer that it looked more like `test/Interop/Cxx/extern-var/extern-var-irgen.swift`.

Comment (MForster) [reply test/Interop/Cxx/operators/non-member-out-of-line-irgen.swift]: How about now?

**Comment by MForster** [comment]
@swift-ci Please test

**Comment by MForster** [comment]
@swift-ci Please test Windows

**Comment by MForster** [comment]
@swift-ci Please test

File: test/Interop/Cxx/operators/non-member-out-of-line-irgen.swift (hunk 2 above)
Comment (gribozavr) [reply test/Interop/Cxx/operators/non-member-out-of-line-irgen.swift]: 👍

**Review by gribozavr — APPROVED** [review]

**Comment by MForster** [comment]
@swift-ci Please test

**Comment by MForster** [comment]
@swift-ci Please test Windows

**Comment by MForster** [comment]
@swift-ci Please smoke test

**Comment by MForster** [comment]
@swift-ci test

**Comment by MForster** [comment]
@swift-ci Please test

**Comment by MForster** [comment]
@swift-ci Please test Linux Platform

**Comment by MForster** [comment]
@swift-ci Please test Windows

**Comment by MForster** [comment]
@swift-ci Please test Linux

**Comment by MForster** [comment]
@swift-ci Please test Windows Platform

**Comment by aschwaighofer** [comment]
The output on armv7 is

```
  %12 = call i32 @_Zpl6IntBoxS_([1 x i32] %6, [1 x i32] %10)
```

arm64 (coercion to i64 happens before that):
```
  %12 = call i64 @_Zpl6IntBoxS_(i64 %6, i64 %10)
```

The quickest way to run these tests on different architectures if you have access to a mac OS X machine is to invoke the build script with `--ios --watchos --reconfigure` and invoke lit directly:

```
.../llvm-project/llvm/utils/lit/lit.py -sv --param swift_test_mode=only_non_executable ./build-master/swift-macosx-x86_64/test-iphoneos-armv7/Interop/Cxx/operators/non-member-out-of-line-irgen.swift
.../llvm-project/llvm/utils/lit/lit.py -sv --param swift_test_mode=only_non_executable ./build-master/swift-macosx-x86_64/test-iphoneos-arm64/Interop/Cxx/operators/non-member-out-of-line-irgen.swift
.../llvm-project/llvm/utils/lit/lit.py -sv --param swift_test_mode=only_non_executable ./build-master/swift-macosx-x86_64/test-watchos-armv7k/Interop/Cxx/operators/non-member-out-of-line-irgen.swift
```

**Comment by MForster** [comment]
Thanks, Arnold. This is very helpful. I was able to fix armv7 before I saw your reply, but arm64 would have been another roundtrip!

**Pull request merged**